src/sparse/impl/KokkosSparse_spmv_impl.hpp

```diff
@@ -577,191 +565,98 @@ struct SPMV_MV_LayoutLeft_Functor {
     // assume either that rows have no duplicate entries, or that rows
     // never have enough duplicate entries to overflow ordinal_type.
 
-#ifdef KOKKOS_ENABLE_PRAGMA_IVDEP
-#pragma ivdep
-#endif
-#ifdef KOKKOS_ENABLE_PRAGMA_UNROLL
-#pragma unroll
-#endif
-#ifdef KOKKOS_ENABLE_PRAGMA_LOOPCOUNT
-#pragma loop count (15)
-#endif
-#ifdef __CUDA_ARCH__
-        for (ordinal_type iEntry = static_cast<ordinal_type> (threadIdx.x);
-             iEntry < row.length;
-             iEntry += static_cast<ordinal_type> (blockDim.x))
-#else
-        for (ordinal_type iEntry = 0;
-             iEntry < row.length;
-             iEntry ++)
-#endif
-      {
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(dev, row.length),
```

Comment:
So I tried to make these changes in the past but I think on some architecture, maybe intel KNL with intel compiler it resulted in a loss of performance and we canned it.
Let's see what @srajama1 says about it now, do we still care about KNL performance or can we make my dream come try and get rid of the ugly pragmas that Kokkos should help us avoid?

Same comment pretty much goes for the rest of the changes in the file

Reply:
@lucbv I'm happy to rerun some benchmarks before & after this change. I saw some comments like "// This should be a thread loop as soon as we can use C++11" so I figured this would be converted to 3-level parallelism at some point and I didn't know performance was a reason to keep it as normal for loops.

I'm pretty sure Kokkos's parallel_for over ThreadVectorRange looks like this for OpenMP:
```
template <typename iType, class Closure, class Member>
KOKKOS_INLINE_FUNCTION void parallel_for(
    Impl::ThreadVectorRangeBoundariesStruct<iType, Member> const&
        loop_boundaries,
    Closure const& closure,
    typename std::enable_if<Impl::is_host_thread_team_member<Member>::value>::
        type const** = nullptr) {
#ifdef KOKKOS_ENABLE_PRAGMA_IVDEP
#pragma ivdep
#endif
  for (iType i = loop_boundaries.start; i < loop_boundaries.end;
       i += loop_boundaries.increment) {
    closure(i);
  }
}
```
so after inlining, I would hope that it basically turns into a regular for loop with ivdep. Then it's basically the original code but without unroll and loop count pragmas.